File: problems/3417-zigzag-grid-traversal-with-skip/solution.py
class Solution:
    def zigzagTraversal(self, grid: List[List[int]]) -> List[int]:
        m = len(grid)
        n = len(grid[0])
        result = []
        row = 0
        left_to_right = True
        while row < m:
            if left_to_right:
                for col in range(0, n, 2):
                    result.append(grid[row][col])
            else:
                for col in range(n - 1, -1, -2):
                    result.append(grid[row][col])
            row += 1
            left_to_right = not left_to_right
        return result

File: problems/3417-zigzag-grid-traversal-with-skip/solution.c
/**
 * Note: The returned array must be malloced, assume caller calls free().
 */
int* zigzagTraversal(int** grid, int gridSize, int* gridColSize, int* returnSize) {
    int m = gridSize;
    int n = gridColSize[0];
    int* result = (int*)malloc(sizeof(int) * ((m * n + 1) / 2));
    *returnSize = 0;
    int row = 0;
    int col = 0;
    int direction = 1; 

    while (row < m) {
        if (direction == 1) {
            int current_col = 0;
            while (current_col < n) {
                result[(*returnSize)++] = grid[row][current_col];
                current_col += 2;
            }
        } else {
            int current_col = n - 1;
            while (current_col >= 0) {
                result[(*returnSize)++] = grid[row][current_col];
                current_col -= 2;
            }
        }
        row++;
        direction *= -1;
    }
    return result;
}